test/Microsoft.AspNet.Cors.Test/CorsMiddlewareTests.cs

```diff
@@ -0,0 +1,217 @@
+                Assert.Equal("AllowedHeader", response.Headers.GetValues(CorsConstants.AccessControlExposeHeaders).FirstOrDefault());
+            }
+        }
+
+        [Fact]
+        public async Task PreFlight_MatchesPolicy_SetsResponseHeaders()
+        {
+            // Arrange
+            var policy = new CorsPolicy();
+            policy.Origins.Add("http://localhost:5001/");
+            policy.Methods.Add("PUT");
+            policy.Headers.Add("Header1");
+            policy.ExposedHeaders.Add("AllowedHeader");
+
+            using (var server = TestServer.Create(app =>
+            {
+                app.UseServices(services =>
+                {
+                    services.AddCors();
+                    services.ConfigureCors(options =>
+                    {
+                        options.AddPolicy("customPolicy", policy);
+                    });
+                });
+                app.UseCors("customPolicy");
+                app.Run(async context =>
+                {
+                    await context.Response.WriteAsync("Cross origin response");
+                });
+            }))
+            {
+                // Act
+                // Preflight request.
+                var response = await server.CreateRequest("/")
+                    .AddHeader(CorsConstants.Origin, "http://localhost:5001/")
+                    .AddHeader(CorsConstants.AccessControlRequestMethod, "PUT")
+                    .SendAsync(CorsConstants.PreflightHttpMethod);
+
+                // Assert
+                response.EnsureSuccessStatusCode();
+                Assert.Equal(2, response.Headers.Count());
+                Assert.Equal("http://localhost:5001/", response.Headers.GetValues(CorsConstants.AccessControlAllowOrigin).FirstOrDefault());
+                Assert.Equal("PUT", response.Headers.GetValues(CorsConstants.AccessControlAllowMethods).FirstOrDefault());
+            }
+        }
+
+        [Fact]
+        public async Task PreFlightRequest_DoesNotMatchPolicy_DoesNotSetHeaders()
+        {
+            // Arrange
+            var policy = new CorsPolicy();
+            policy.Origins.Add("http://localhost:5001/");
+            policy.Methods.Add("PUT");
+            policy.Headers.Add("Header1");
+            policy.ExposedHeaders.Add("AllowedHeader");
+
+            using (var server = TestServer.Create(app =>
+            {
+                app.UseServices(services => services.AddCors());
+                app.UseCors(policy);
+                app.Run(async context =>
+                {
+                    await context.Response.WriteAsync("Cross origin response");
+                });
+            }))
+            {
+                // Act
+                // Preflight request.
+                var response = await server.CreateRequest("/")
+                    .AddHeader(CorsConstants.Origin, "http://localhost:5001/sub")
```

Comment:
`http://localhost:5001/sub` can't be an origin. In a real world, no user agent will sent a preflight request like this. An origin is a combination of URI scheme, hostname, and port number. http://www.w3.org/Security/wiki/Same_Origin_Policy